tools/c7n_azure/tests/test_session.py

```diff
@@ -22,6 +27,89 @@ class SessionTest(BaseTest):
     def setUp(self):
         super(SessionTest, self).setUp()
 
+    def mock_init(self, client_id, secret, tenant, resource):
+        pass
+
+    def test_initialize_session_principal(self):
+        with patch('azure.common.credentials.ServicePrincipalCredentials.__init__', autospec=True, return_value=None):
+            with patch.dict(os.environ,
+                            {
+                                'AZURE_TENANT_ID': 'tenant',
+                                'AZURE_SUBSCRIPTION_ID': 'ea42f556-5106-4743-99b0-c129bfa71a47',
+                                'AZURE_CLIENT_ID': 'client',
+                                'AZURE_CLIENT_SECRET': 'secret'
+                            }, clear=True):
+
+                s = Session()
+                s._initialize_session()
+                self.assertIs(type(s.credentials), ServicePrincipalCredentials)
+                self.assertEqual(s.subscription_id, 'ea42f556-5106-4743-99b0-c129bfa71a47')
```

Comment:
you can use s.get_subscription_id() & s.get_credentials().. (It should also call _initialize_session)
(Same in other tests)